Link Visual and Collision are not exported to XML

AdrianZw opened the pull request:
I realized, that Visual and Collision are not exported to XML, if they are set via the Constructor.
There is also an open issue for it #42.
In addition, I realized that this issue was already fixed on the melodic-branch, see pull request #47.

Consequently, I cherry-pick the changes introduced with #47. In addition, I had to remove one test, since the `Visual` class does not have an attribute `name` anymore.


Closes: #42
Relates: #47

k-okada committed:
update backward compatibility on visual and collisions (#47)

* add test to build robot model from python urdf classes

* add more tests on python urdf build

* update backward compatibility

Link(visual=...) does not set <visual> on <link> since 0.4.0, breaks backward compatibility

* Robot accepts version kwarg

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>
Co-authored-by: Shane Loretz <sloretz@osrfoundation.org>

AdrianZw committed:
Remove test for visual with name

The Visual class does not have an attribute name anymore. Consequently,
we can not expect the test to work.

clalancette reviewed (changes requested):
Can you please retarget these changes to the `ros2` branch?  That's where most of the development for ROS 2 happens.  Once that is done, I'll do a more detailed review.  Thanks!

AdrianZw commented:
> Can you please retarget these changes to the `ros2` branch? That's where most of the development for ROS 2 happens. Once that is done, I'll do a more detailed review. Thanks!

I did not find a way to change the branch from which the changes are pulled from. (Unless opening a new PR) Thus, I repurposed the `foxy-devel` branch. Might not be the cleanest way, to do it ...

clalancette commented on src/urdf_parser_py/urdf.py:
```diff
@@ -370,6 +374,8 @@ def __set_visual(self, visual):
         else:
             self.visuals.append(visual)
+        if visual:
+            self.add_aggregate('visual', visual)
```
So this isn't totally the fault of this PR, but there definitely seems to be some confusion in here around `visuals`/`visual` and `collisions`/`collision`.  At least, I am confused.

What I *think* is intended here is that a link can contain multiple `visual` and `collision` tags.  And the `visuals` and `collisions` list is supposed to hold that.  But there is also a `visual` and `collision` property for back-compat reasons.

But I think those back-compat versions should really only work properly for a link with a single one of them.  So I actually think this should be something more like the following:

```
    def __set_visual(self, visual):
        """Set the first visual."""
        if self.visuals:
            self.remove_aggregate(self.visuals[0])
        if visual:
            self.add_aggregate('visual', visual)
```

That ensures that only the "last" visual set is the one that is preserved.

This then begs the question of what we do when we want multiple `visual` or `collision` tags.  In that case, I think we need to add in a new method which is something like `add_visual` or `add_collision`, which would strictly append to the list.

Does this all make sense to you?

AdrianZw replied:
Yes, it makes sense to me. I added `add_visual` and `add_collision` methods and refactored the unit tests.

Do we also need `remove_visual(visual)` or `remove_collision(collision)` methods?

Let me know what you think :smile:

clalancette replied:
> Do we also need `remove_visual(visual)` or `remove_collision(collision)` methods?

Yeah, I think that would be a good idea.  Thanks!

AdrianZw replied:
Actually, it was a good thing to add these methods, since in the previous version I added the objects twice...

It is actually sufficient to simply call `add_aggregate` or `remove_aggregate`. There is no need to manipulate the lists `Link.collisions` or  `Link.visuals` directly.

AdrianZw committed:
Add method for multiple visual and collision props

The attributes 'visual' and 'collision' exist only for back-compability reasons
and shall only be used if there exists one and only one visual or collision element.
For links with multiple visual or collision elements the new method is added which
shall be used. This simplifies the usage of the API.

AdrianZw committed:
Add methods to remove visuall and collision objs

- In the previous version, collision and visual objects were
added twice. Tests were updated accordingly.
- Added tests to check if visual and collision objects can be removed.
